// @hubspot/local-dev-lib/enums/build.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SUBCOMPONENT_TYPES = exports.COMPONENT_TYPES = exports.DEPLOYABLE_STATES = exports.SUBBUILD_TYPES = exports.BUILD_STATUS = void 0;
exports.BUILD_STATUS = {
    BUILDING: 'BUILDING',
    FAILURE: 'FAILURE',
    PREPARING: 'PREPARING',
    PENDING: 'PENDING',
    SUCCESS: 'SUCCESS',
};
exports.SUBBUILD_TYPES = {
    APP: 'APP',
    PRIVATE_APP: 'PRIVATE_APP',
    THEME: 'THEME',
    REACT_THEME: 'REACT_THEME',
    NO_SUBBUILDS: 'NO_SUBBUILDS',
};
exports.DEPLOYABLE_STATES = {
    DEPLOYABLE: 'DEPLOYABLE',
    DEPRECATED: 'DEPRECATED_APP_NON_SUBCOMPONENT',
};
exports.COMPONENT_TYPES = {
    APP: 'APP',
    PRIVATE_APP: 'PRIVATE_APP',
    THEME: 'THEME',
    REACT_THEME: 'REACT_THEME',
};
exports.SUBCOMPONENT_TYPES = {
    APP_ID: 'APP_ID',
    PACKAGE_LOCK_FILE: 'PACKAGE_LOCK_FILE',
    CRM_CARD_V2: 'CRM_CARD_V2',
    CARD_V2: 'CARD_V2',
    SERVERLESS_PKG: 'SERVERLESS_PKG',
    SERVERLESS_ROUTE: 'SERVERLESS_ROUTE',
    SERVERLESS_FUNCTION: 'SERVERLESS_FUNCTION',
    APP_FUNCTION: 'APP_FUNCTION',
    AUTOMATION_ACTION: 'AUTOMATION_ACTION',
    REACT_EXTENSION: 'REACT_EXTENSION',
};
